### PR TITLE
[clang][bytecode] Implement comparsion operators for vector type

### PR DESCRIPTION
**Pull request by yronglin**
Implement ==, !=, <, <=, >, >= comparsion operators for vector type.

### CODE REVIEW

File: clang/lib/AST/ByteCode/Compiler.cpp [hunk 1]
```diff
@@ -1222,6 +1224,117 @@ bool Compiler<Emitter>::VisitComplexBinOp(const BinaryOperator *E) {
   return true;
 }
 
+template <class Emitter>
+bool Compiler<Emitter>::VisitVectorBinOp(const BinaryOperator *E) {
+  assert(E->getType()->isVectorType());
```
Comment (tbaederr) [inline clang/lib/AST/ByteCode/Compiler.cpp:1229]: Add an assertion that both LHS and RHS are also of vector type.

File: clang/lib/AST/ByteCode/Compiler.cpp [hunk 2]
```diff
@@ -1222,6 +1224,117 @@ bool Compiler<Emitter>::VisitComplexBinOp(const BinaryOperator *E) {
...
+      llvm_unreachable("Unsupported binary operator");
+    }
+    if (!this->emitInitElem(ResultElemT, I, E))
+      return false;
+  }
+  return true;
+}
+
+template <class Emitter>
+bool Compiler<Emitter>::emitVectorComparison(const BinaryOperator *E) {
```
Comment (tbaederr) [inline clang/lib/AST/ByteCode/Compiler.cpp:1289]: I don't think we're going to use this anywhere else, so a separate function shouldn't be necessary

File: clang/lib/AST/ByteCode/Compiler.cpp [hunk 3]
```diff
@@ -1222,6 +1224,117 @@ bool Compiler<Emitter>::VisitComplexBinOp(const BinaryOperator *E) {
...
+    unsigned LocalIndex = allocateTemporary(E);
+    if (!this->emitGetPtrLocal(LocalIndex, E))
+      return false;
+  }
+
+  const Expr *LHS = E->getLHS();
+  const Expr *RHS = E->getRHS();
+  const auto *VecTy = E->getType()->getAs<VectorType>();
+
+  // The LHS and RHS of a comparison operator must have the same type. So we
```
Comment (tbaederr) [inline clang/lib/AST/ByteCode/Compiler.cpp:1246]: Add an assertion if that's the case.

File: clang/lib/AST/ByteCode/Compiler.cpp (hunk 1 above)
Comment (yronglin) [reply clang/lib/AST/ByteCode/Compiler.cpp:1229]: Done.

File: clang/lib/AST/ByteCode/Compiler.cpp (hunk 3 above)
Comment (yronglin) [reply clang/lib/AST/ByteCode/Compiler.cpp:1248]: Done.

File: clang/lib/AST/ByteCode/Compiler.cpp (hunk 2 above)
Comment (yronglin) [reply clang/lib/AST/ByteCode/Compiler.cpp]: Agree, removed.

File: clang/lib/AST/ByteCode/Compiler.cpp [hunk 4]
```diff
@@ -1222,6 +1224,105 @@ bool Compiler<Emitter>::VisitComplexBinOp(const BinaryOperator *E) {
...
+    return false;
+
+  // Evaluate RHS and save value to RHSOffset.
+  unsigned RHSOffset = this->allocateLocalPrimitive(RHS, PT_Ptr, true, false);
+  if (!this->visit(RHS))
+    return false;
+  if (!this->emitSetLocal(PT_Ptr, RHSOffset, E))
+    return false;
+
+  auto getElem = [=](unsigned Offset, unsigned Index, PrimType ElemT) -> bool {
```
Comment (tbaederr) [inline clang/lib/AST/ByteCode/Compiler.cpp:1267]: `ElemT` can be removed since it's always the same anyway

Comment (yronglin) [reply clang/lib/AST/ByteCode/Compiler.cpp]: Thanks, removed!

**Review by tbaederr — APPROVED** [review]

**Comment by yronglin** [comment]
Thanks for the review, wait for CI green before land.

**Pull request merged**